plat-stm32mp1: IO domains updates

etienne-lms opened the pull request:
This series implements STM32MP1 variants IO domain feature.
- IO compensation for STM32MP13 variant
- High Speed Low Voltage for STM32MP13 and STM32MP15
- IO domain aware regulators for STM32MP13 variants

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

patrickdelaunay commented on core/arch/arm/dts/stm32mp131.dtsi:
```diff
@@ -83,6 +83,16 @@
 
+	sdmmc1_io: sdmmc1_io {
+		compatible = "st,stm32mp13-iod";
+		regulator-name = "sdmmc1_io";
```
+		regulator-always-on;

you need  to define the IO domain as always on regulator in STM32MP13 SoC dtsi, to enable
IOCOMP  and  deactivate the level detectors in SYSCFG

in most of case the input regulator can't be stopeed (only allow 3V3 to 1V8 switch).

For "always-on" regulator, it is enabled during the IOD regulator probe 
It will be never deactivated by regulator framework (it is no required by Linux = VQMMC)  
and it is no issue to keep this feature enabled.

The ICOMP init is no more done in SYSCFG for IOD managed in driver stm32mp1_regulator_iod.c 
see test  on SYSCFG_IO_COMP_IDX_SD1 / SYSCFG_IO_COMP_IDX_SD2 in syscfg driver

it is done in SYSCFG driver  ONLY for IOCOMP linked to VDD.

etienne-lms replied:
Right, thanks. I confused the SDMMC device regulator supply and the SDMMC IO domain regulator supply.

patrickdelaunay commented on core/arch/arm/dts/stm32mp131.dtsi:
```diff
@@ -83,6 +83,16 @@
+
+	sdmmc2_io: sdmmc2_io {
+		compatible = "st,stm32mp13-iod";
+		regulator-name = "sdmmc2_io";
```
+		regulator-always-on;

patrickdelaunay commented on core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c:
Reviewed-by: Patrick Delaunay <patrick.delaunay@foss.st.com>

patrickdelaunay reviewed (commented):
correct address for upstream :  <p.paillet@st.com> => <p.paillet@foss.st.com>

etienne-lms commented:
Added `regulator-always-on` property to the 2 nodes.
Fixed Pascal's e-mail address in related commit messages.
Applied Patrick's review tag.

GseoC commented on core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.h:
```diff
@@ -0,0 +1,51 @@
+void stm32mp_set_vddsd_comp_state(enum stm32mp13_vddsd_comp_id id, bool enable);
+
+/*
+ * Enable or disable High Speed Low Voltage mmode of an IO domain
```
typo: mmode/mode

GseoC commented on core/drivers/regulator/stm32mp1_regulator_iod.c:
```diff
@@ -0,0 +1,296 @@
+
+#define TIMEOUT_US_10MS		U(10000)
+
+#define IO_VOLTAGE_THRESHOLD_UV	2700000
```
minor: U(2700000)

etienne-lms replied:
Voltage levels are of `int` type hence signed.

GseoC commented on core/drivers/regulator/stm32mp1_regulator_iod.c:
```diff
@@ -0,0 +1,296 @@
+
+struct regulator *stm32mp1_get_iod_regulator(enum iod_regulator_id index)
+{
+	assert(index > 0 && index < IOD_REGU_COUNT);
```
isn't IOD_SDMMC1=0?

etienne-lms replied:
Right, thanks.
Looks like i missed to test that in debug mode. Done now, much better with the fixup!

GseoC commented on core/drivers/regulator/stm32mp1_regulator_iod.c:
```diff
@@ -0,0 +1,296 @@
+}
+
+/*
+ * To protect the IOs, we switch disable High Speed Low Voltage mode before
```
s/we switch disable/we disable

GseoC commented on core/drivers/regulator/stm32mp1_regulator_iod.c:
```diff
@@ -0,0 +1,296 @@
+
+	FMSG("iod probe node '%s'", node_name);
+
+	/* Look up matching regulator name defein in SoC DTSI file */
```
s/defein/defined

etienne-lms commented:
comments addressed.

etienne-lms commented:
@GseoC, have I successfully addressed your comments?

GseoC commented on core/drivers/regulator/stm32mp1_regulator_iod.c:
```diff
@@ -0,0 +1,296 @@
+};
+
+DEFINE_DT_DRIVER(stm32mp1_regulator_iod_dt_driver) = {
+	.name = "stm32mp1-iod-regulator",
```
This driver is only supported on STM32MP13x.

Shall we rename it or are we anticipating changes on STM32MP15?

GseoC replied:
> STM32MP15 does have have needs runtime voltage level re-configurable IO domains. I'll rename the driver to "stm32mp13-iod-regulator".

I guess the file name itself can be changed as well

etienne-lms replied:
make sense, i'll do.

GseoC commented on core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c:
```diff
@@ -16,87 +18,175 @@
+	hslvenxr_offset = SYSCFG_HSLVEN0R + id * sizeof(uint32_t);
 
-	io_write32(syscfg_base + SYSCFG_CMPCR, value | SYSCFG_CMPCR_SW_CTRL);
+	io_write32(get_syscfg_base() + hslvenxr_offset, hlvs_value);
```
We  could read back if High speed mode is enabled by reading 1 to this register

`
	if (enable && !(io_write32(get_syscfg_base() + hslvenxr_offset) & 0x1))
		panic();
	else if (!enable && io_write32(get_syscfg_base() + hslvenxr_offset) )
		panic();
`

GseoC reviewed (commented):
A few comments, the rest looks ok

etienne-lms commented:
Comments addressed.
I've rebased the series on master tip.

GseoC commented on core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.c:
```diff
@@ -178,13 +181,81 @@ struct regulator *stm32mp1_pwr_get_regulator(enum pwr_regulator id)
+					      "stm32mp1-pwr-hslv");
+		}
+	} else if (product_below_2v5) {
+		panic("Vdd to high for related IO domains");
```
s/to/too

GseoC commented on core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.h:
```diff
@@ -17,6 +17,14 @@
 #define PWR_WKUPCR_OFF		0x20
 #define PWR_MPUWKUPENR_OFF	0x28
 
+/* CR3 register bitfield for STM432MP13 variants */
```
No STM432 yet 😄

GseoC commented on core/drivers/regulator/stm32mp1_regulator_iod.c:
```diff
@@ -0,0 +1,296 @@
+	enum stm32mp13_hslv_id hslv_id;
+	enum stm32mp13_vddsd_comp_id io_comp_id;
+	bool suspend_state;
+	int suspend_level_uv;
```
nitpick: I'd move suspend_level_uv above for padding

etienne-lms replied:
`bool` is the size of an `int` in current builds.
I prefer as-is.

GseoC commented:
+
> > STM32MP15 does have have needs runtime voltage level re-configurable IO domains. I'll rename the driver to "stm32mp13-iod-regulator".
> 
> I guess the file name itself can be changed as well

With comment addressed:
`Acked-by: Gatien Chevallier <gatien.chevallier@foss.st.com>`

etienne-lms commented:
Review comments addressed. As I renamed the stm32mp13 IO domain regulator driver source file, I also changed the header file name, the driver name and the related config switch (s/CFG_STM32MP1_REGULATOR_IOD/CFG_STM32MP13_REGULATOR_IOD/g).

Let's run CI tests. I'll squashed fixup commits and apply your review tag afterward.

etienne-lms commented:
Thanks for the review.
Review tags applied.
